timemachine/cpp/src/kernels/k_barostat.cuh

```diff
@@ -0,0 +1,153 @@
+#pragma once
+
+#include "k_fixed_point.cuh"
+
+void __global__ rescale_positions(
+    const int N,                                     // Number of atoms to shift
+    double *__restrict__ coords,                     // Coordinates
+    const double *__restrict__ length_scale,         // [1]
+    const double *__restrict__ box,                  // [9]
+    double *__restrict__ scaled_box,                 // [9]
+    const int *__restrict__ atom_idxs,               // [N]
+    const int *__restrict__ mol_idxs,                // [N]
+    const int *__restrict__ mol_offsets,             // [N]
+    const unsigned long long *__restrict__ centroids // [N*3]
+) {
+    int idx = blockIdx.x * blockDim.x + threadIdx.x;
+    float center_x = box[0 * 3 + 0] * 0.5f;
+    float center_y = box[1 * 3 + 1] * 0.5f;
+    float center_z = box[2 * 3 + 2] * 0.5f;
+
+    double scale = length_scale[0];
+    if (idx == 0) {
+        scaled_box[0 * 3 + 0] *= scale;
+        scaled_box[1 * 3 + 1] *= scale;
+        scaled_box[2 * 3 + 2] *= scale;
+    }
+    while (idx < N) {
+        int atom_idx = atom_idxs[idx];
+        int mol_idx = mol_idxs[idx];
+
+        float num_atoms = static_cast<float>(mol_offsets[mol_idx + 1] - mol_offsets[mol_idx]);
+
+        float centroid_x = FIXED_TO_FLOAT<float>(centroids[mol_idx * 3 + 0]) / num_atoms;
+        float centroid_y = FIXED_TO_FLOAT<float>(centroids[mol_idx * 3 + 1]) / num_atoms;
+        float centroid_z = FIXED_TO_FLOAT<float>(centroids[mol_idx * 3 + 2]) / num_atoms;
+
+        float displacement_x = ((centroid_x - center_x) * scale) + center_x - centroid_x;
+        float displacement_y = ((centroid_y - center_y) * scale) + center_y - centroid_y;
+        float displacement_z = ((centroid_z - center_z) * scale) + center_z - centroid_z;
+
+        coords[atom_idx * 3 + 0] += displacement_x;
+        coords[atom_idx * 3 + 1] += displacement_y;
+        coords[atom_idx * 3 + 2] += displacement_z;
+
+        idx += gridDim.x * blockDim.x;
+    }
+}
+
+void __global__ find_group_centroids(
+    const int N,                               // Number of atoms to shift
+    const double *__restrict__ coords,         // Coordinates
+    const int *__restrict__ atom_idxs,         // [N]
+    const int *__restrict__ mol_idxs,          // [N]
+    unsigned long long *__restrict__ centroids // [num_molecules * 3]
+) {
+    int idx = blockIdx.x * blockDim.x + threadIdx.x;
+    while (idx < N) {
```

Comment:
question: why is the `while` loop needed here? (aren't we already requesting one thread per atom?)

Reply:
I added that so that we could play around with different numbers of blocks. Currently, we launch as many blocks as we need, but as the system sizes get larger this may not be optimal.

I see us writing kernels like this more broadly to be able to adjust the number of blocks in a more flexible, as well as to have threads do more work.

Reply:
Makes sense.

(seems like there could be a _tiny_ overhead from the unnecessary `idx` increment, but doesn't seem likely to be noticeable..)

Would it be worth clarifying this with a comment to avoid any future confusion?

Reply:
Don't know that a comment is very useful as it is a pretty common pattern and one we already use in the nonbonded potential: https://github.com/proteneer/timemachine/blob/master/timemachine/cpp/src/kernels/k_nonbonded.cuh#L383

Reply:
Hmm, ok. This case seems different to me because it seems like we'll never do more than a single iteration given the way the kernel is invoked. But I may be misunderstanding something.

Reply:
As it is called currently, that is true. Will add a comment in the calling code that the number of blocks could be fewer than N / tpb

Reply:
Added a comment in https://github.com/proteneer/timemachine/pull/1101/commits/e457f249da32d03b6359e78a176ada0b21063114